### DISCUSSION TITLE
Episode 115: Max Gap in C++23

### DISCUSSION THREAD

**Discussion by codereport** [codereport/adsp2#5] [Episode]
Episode Link: https://adspthepodcast.com/2023/02/03/Episode-115.html

Feel free to write comments or ask questions :slightly_smiling_face:

**Comment by gummif** [comment]
The max gap problem states that it must be solved in linear time but the solution uses sort which is N log N.

**Reply by codereport** [reply]
> The max gap problem states that it must be solved in linear time but the solution uses sort which is N log N.

Yea, I noticed this as well. For the purposes of our discussion it didn't change anything. The argument in the discussion is that if you use a Radix Sort that it can become `O(kn)` (see the [LeetCode discussion section](https://leetcode.com/problems/maximum-gap/discussion/)). I think it shouldn't be labeled "hard though" because an an `O(nlogn)` works completed fine for the constraints of the problem. 

![image](https://user-images.githubusercontent.com/36027403/218180587-4c524bec-1463-4064-bfa4-95748da40be7.png)

**Reply by gummif** [reply]
Yes definitely more elegent with sort and probably similar performance.

**Comment by DanielDewberry** [comment]
Hello. Thankyou for another interesting episode. Please could you share a url to the optimization book mentioned?

**Reply by codereport** [reply]
@brycelelbach Another question for you : )

**Reply by brycelelbach** [reply]
[An Introduction to Optimization by Edwin K. P. Chong and Stanislaw H. Zak (Author)](https://www.amazon.com/Introduction-Optimization-Edwin-K-Chong/dp/1118279018). It's expensive though; there's probably better and cheaper options out there.

**Reply by DanielDewberry** [reply]
Thank you @brycelelbach. I'll add it to my ever growing reading list :)

**Reply by joaogui1** [reply]
Could you add the link to the episode's links?

**Reply by codereport** [reply]
Done (note that I added it to ep 116 because that is when it was mentioned - even though this is the 115 discussion)

**Comment by kerron** [comment]
Here's a Haskell solution to the problem mentioned in Ep.115 and the follow up in Ep.116

```hs
sumList :: [Int] -> [Int]
sumList xs =  [ y - x | (x, y) <- zip xs (tail xs)]

-- Ep.115
maxGap :: [Int] -> Int
maxGap xs = maximum (sumList xs)

-- Ep.116
maxCount :: [Int] -> Int
maxCount xs = length (filter (== maxGap xs) (sumList xs))
```

**Reply by codereport** [reply]
I love Haskell ❤️ My solutions would be:

```hs
import Data.List.HT (mapAdjacent)
import Data.List.Utils (countElem)

-- Ep.115
maxGap :: [Int] -> Int
maxGap xs =  maximum . mapAdjacent (flip (-))

-- Ep.116
maxCount :: [Int] -> Int
maxCount xs = (flip countElem) <*> maximum $ mapAdjacent (flip (-))
```
`flip` is the C combinator and `<*>` is the S combinator 🙂

**Comment by waydan** [comment]
@brycelelbach, I'm catching up on episodes. Did you ever find a Japanese speaker to help you get your recipe?

**Reply by codereport** [reply]
Last thing I heard he did not 😢

**Reply by waydan** [reply]
I happen to speak Japanese and could possibly help if Bryce still needs it.
